### PR TITLE
RFC: Add design for blueprint and phase progress calculation

### PR DESCRIPTION
**Pull request by ihcsim**
## Change Overview

This PR adds the design proposal for the blueprint and phase progress calculation. Welcome all comments and feedback.

## Pull request type

Please check the type of change your PR introduces:
- RFC

## Issues

- #1455

### CODE REVIEW

File: design/progress-tracking.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+* `ExportRDSSnapshotToLocation`
+* `RestoreRDSSnapshot`
+
+#### Blueprint Progress
+
+Initially, the blueprint overall progress is computed by checking the number of
+completed phases against the total number of phases:
+
+```
+blueprint_progress_percent = num_completed_phases / total_num_phases * 100
```
Comment (viveksinghggits) [inline design/progress-tracking.md:91]: should this be `action_progress_percent`?

File: design/progress-tracking.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+* parsing the log outputs for milestones indicator
+
+✍️ Currently, Kanister Functions do not utilize Kopia to perform their
+underlying work. Once the work to integrate Kopia into Kanister is completed,
+we can extract the progress status directly from the log outputs.
+
+Here's a sample log output of the Kopia create snapshot function:
+
+```sh
+$ kopia snapshot create kanister
```
Comment (viveksinghggits) [inline design/progress-tracking.md:127]: In both the examples (kopia and restic) we are just considering the utilities that are going to upload the data to object store but in some cases there are chances that there are some other commands that are taking some time.
- One example is the utility that we are using to take backup for example `mysqldump`, `esdump` etc. Should we consider these as well or just measuring upload time would be sufficient.

Comment (dsu-igeek) [reply design/progress-tracking.md:127]: Definitely if we can get progress from the tools we should expose that.

**Review by dsu-igeek — APPROVED** [review]
I like this.  Thanks for putting this together!

File: design/progress-tracking.md (hunk 1 above)
Comment (ihcsim) [reply design/progress-tracking.md:91]: @viveksinghggits yes, good catch! To get the blueprint overall progress, we should be assessing the completed actions, not phases.

File: design/progress-tracking.md (hunk 2 above)
Comment (ihcsim) [reply design/progress-tracking.md:127]: @viveksinghggits Yes, the goal is that downstream users can add more custom progress trackers (at the code level) per their requirements, if the built-in ones don't work. We can consider adding a built-in generic tracker to output to 25%, 50%, 75% and 100% markers based on certain events in the logs.

File: design/progress-tracking.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+### Assumptions And Constraints
+
+* The progress computation should not compromise the main data protection task's
+latency nor lead to resource contention. Progress computation will be performed
+on a best-effort basis, where it may be de-prioritized with no guarantee on the
+accuracy of its result, or skipped entirely in the event of resource contention.
+
+### Changes To API
+
+The blueprint's overall progress will be reported under the new
```
Comment (pavannd1) [inline design/progress-tracking.md:35]: Similar to Vivek's suggestion, should this say action/actionset's overall progress too?

File: design/progress-tracking.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+* `ExportRDSSnapshotToLocation`
+* `RestoreRDSSnapshot`
+
+#### Blueprint Progress
+
+Initially, the blueprint overall progress is computed by checking the number of
+completed actions against the total number of actions:
+
+```
+blueprint_progress_percent = num_completed_actions / total_num_actions * 100
```
Comment (pavannd1) [inline design/progress-tracking.md:91]: s/blueprint/actionset here and below?

File: design/progress-tracking.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+#### Error Handling
+
+If a phase failed, the progress tracking will cease immediately. The last
+reported progress will be retained in the actionset's `status` subresource.
+
+## Test Cases
+
+New unit tests to be added to the new `progress` package to cover blueprint
+progress, and phase progress calculated with a dummy tracker:
```
Comment (pavannd1) [inline design/progress-tracking.md:238]:
```suggestion
progress, and phase progress calculated with a sample tracker:
```
😯 https://twitter.com/TwitterEng/status/1278733305190342656/photo/1

File: design/progress-tracking.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    percentCompleted: 50 # 1 out of 2 phases are completed
+    lastTransitionTime: 2022-04-06 14:23:34
+  actions:
+  - blueprint: my-blueprint
+    name: action-00
+    phases:
+    - name: echo
+      state: completed
+      progress:
+        percentCompleted: 100
```
[the remark this answers is not in the document]
Comment (ihcsim) [reply design/progress-tracking.md]: Turns out that we may have to use integers after all, due to the possibility of precision loss on round-trip conversion. See https://github.com/kubernetes-sigs/controller-tools/issues/245.

[the remark this answers is not in the document]
Comment (ihcsim) [reply design/progress-tracking.md]: If the fractional part is important, we will need to convert the whole float into an integer before persisting them. But the caller will need to convert it back to float.

**Review by pavannd1 — APPROVED** [review]

**Pull request merged**